### PR TITLE
Readonly classes cannot be combined with `#[AllowDynamicProperties]`

### PR DESCRIPTION
**Pull request by staabm**
closes https://github.com/phpstan/phpstan/issues/12281

### CODE REVIEW

File: tests/PHPStan/Rules/Classes/data/bug-12281.php [hunk 1]
```diff
@@ -0,0 +1,10 @@
+<?php // lint >= 8.2
+
+namespace Bug12281;
+
+#[\AllowDynamicProperties]
+readonly class BlogData { /* … */ }
+
+/** @readonly  */
+#[\AllowDynamicProperties]
+class BlogDataPhpdoc { /* … */ }
```
Comment (staabm) [inline tests/PHPStan/Rules/Classes/data/bug-12281.php:10]: not sure we want the same error for phpdoc readonly, as we have for native readonly...?

looking thru the source we usually only do the "readonly-by-phpdoc" dance for properties, not for classes

File: src/Rules/Classes/ClassAttributesRule.php [hunk 2]
```diff
@@ -28,12 +31,23 @@ public function processNode(Node $node, Scope $scope): array
...
 			$scope,
 			$classLikeNode->attrGroups,
 			Attribute::TARGET_CLASS,
 			'class',
 		);
+
+		$classReflection = $scope->getClassReflection();
+		if ($classReflection !== null && $classReflection->isReadOnly()) {
+			if (count($classReflection->getNativeReflection()->getAttributes('AllowDynamicProperties')) > 0) {
+				$errors[] = RuleErrorBuilder::message(sprintf('Cannot apply #[AllowDynamicProperties] to readonly class %s.', $classReflection->getName()))
```
Comment (ondrejmirtes) [inline src/Rules/Classes/ClassAttributesRule.php]: We don't use this error message format anywhere. I'd follow some error from AttributesCheck. Example:

> Attribute class AllowDynamicProperties cannot be used with readonly classes.

File: src/Rules/Classes/ClassAttributesRule.php [hunk 3]
```diff
@@ -28,12 +31,23 @@ public function processNode(Node $node, Scope $scope): array
...
+					->identifier('class.allowDynamicPropertiesReadonly')
+					->build();
```
Comment (ondrejmirtes) [inline src/Rules/Classes/ClassAttributesRule.php:62]: Should be non-ignorable

File: tests/PHPStan/Rules/Classes/data/bug-12281.php (hunk 1 above)
Comment (ondrejmirtes) [reply tests/PHPStan/Rules/Classes/data/bug-12281.php:10]: I wouldn't do that for PHPDoc `@readonly`.

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]
We also need to error when this attribute is used above interfaces. Also please check what happens above enums and traits (I'm not sure).

**Comment by staabm** [comment]
> We also need to error when this attribute is used above interfaces. Also please check what happens above enums and traits (I'm not sure).

I have tested this cases and linked the 3v4l.org examples in https://github.com/phpstan/phpstan/issues/12281#issue-2744490661

TL;DR:
- enums not allowed
- interfaces not allowed
- traits not allowed

what do you think about adding a separate `AllowDynamicPropertiesAttributeRule` which targets `ClassLike` instead of adding this error to the existing rule?
(atm we don't have a rule which covers attributes on traits)

File: src/Rules/Classes/ClassAttributesRule.php [hunk 4]
```diff
@@ -28,12 +31,34 @@ public function processNode(Node $node, Scope $scope): array
...
 
-		return $this->attributesCheck->check(
+		$errors = $this->attributesCheck->check(
 			$scope,
 			$classLikeNode->attrGroups,
 			Attribute::TARGET_CLASS,
 			'class',
 		);
+
+		$classReflection = $scope->getClassReflection();
```
Comment (ondrejmirtes) [inline src/Rules/Classes/ClassAttributesRule.php]: You can ask `$node` for the class reflection, it's not nullable on that.

File: src/Rules/Classes/ClassAttributesRule.php [hunk 5]
```diff
@@ -28,12 +31,34 @@ public function processNode(Node $node, Scope $scope): array
...
+		if (
+			$classReflection !== null
+			&& ($classReflection->isReadOnly() || $classReflection->isEnum() || $classReflection->isInterface())
+		) {
+			$message = 'Attribute class AllowDynamicProperties cannot be used with readonly classes.';
+			if ($classReflection->isEnum()) {
+				$message = 'Attribute class AllowDynamicProperties cannot be used with enums.';
+			}
+			if ($classReflection->isInterface()) {
+				$message = 'Attribute class AllowDynamicProperties cannot be used with interface.';
```
Comment (ondrejmirtes) [inline src/Rules/Classes/ClassAttributesRule.php]: This could by DRYed and cleaned up with sprintf.

File: src/Rules/Classes/ClassAttributesRule.php [hunk 6]
```diff
@@ -28,12 +31,34 @@ public function processNode(Node $node, Scope $scope): array
...
+			}
+
+			if (count($classReflection->getNativeReflection()->getAttributes('AllowDynamicProperties')) > 0) {
+				$errors[] = RuleErrorBuilder::message($message)
+					->identifier('class.allowDynamicPropertiesReadonly')
```
Comment (ondrejmirtes) [inline src/Rules/Classes/ClassAttributesRule.php]: Needs better identifier for enums and interfaces.

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]
Also:

1) Please do new TraitAttributesRule similar to ClassAttributesRule that will call AttributesCheck. We need the rule to hook onto Trait_ and fetch the reflection ourselves from ReflectionProvider.
2) Add AllowDynamicProperties check there too.

File: tests/PHPStan/Rules/Classes/data/bug-12281.php [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+
+#[\AllowDynamicProperties]
+enum BlogDataEnum { /* … */ }
+
+#[\AllowDynamicProperties]
+interface BlogDataInterface { /* … */ }
+
+#[\AllowDynamicProperties]
+trait BlogDataTrait { /* … */ }
```
Comment (ondrejmirtes) [inline tests/PHPStan/Rules/Classes/data/bug-12281.php]: This test is irrelevant for this rule so move it elsewhere.

File: src/Rules/Classes/ClassAttributesRule.php [hunk 8]
```diff
@@ -28,12 +31,39 @@ public function processNode(Node $node, Scope $scope): array
...
+		if (
+			$classReflection->isReadOnly()
+			|| $classReflection->isEnum()
+			|| $classReflection->isInterface()
+		) {
+			$typeName = 'readonly class';
+			$identifier = 'class.allowDynamicPropertiesReadonly';
+			if ($classReflection->isEnum()) {
+				$typeName = 'enum';
+				$identifier = 'class.allowDynamicPropertiesEnum';
```
Comment (ondrejmirtes) [inline src/Rules/Classes/ClassAttributesRule.php]: enum.allowDynamicProperties

File: src/Rules/Classes/ClassAttributesRule.php [hunk 9]
```diff
@@ -28,12 +31,39 @@ public function processNode(Node $node, Scope $scope): array
...
+			}
+			if ($classReflection->isInterface()) {
+				$typeName = 'interface';
+				$identifier = 'class.allowDynamicPropertiesInterface';
```
Comment (ondrejmirtes) [inline src/Rules/Classes/ClassAttributesRule.php]: interface.allowDynamicProperties

File: src/Rules/Traits/TraitAttributesRule.php [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+		}
+
+		if (!$this->reflectionProvider->hasClass($traitName->toString())) {
+			return [];
+		}
+
+		$errors = $this->attributesCheck->check(
+			$scope,
+			$node->attrGroups,
+			Attribute::TARGET_ALL, // there is no dedicated trait target
```
Comment (ondrejmirtes) [inline src/Rules/Traits/TraitAttributesRule.php]: TARGET_CLASS should be used https://3v4l.org/DJPUf#v8.4.2

File: src/Rules/Traits/TraitAttributesRule.php [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+		if ($traitName === null) {
+			return [];
+		}
+
+		if (!$this->reflectionProvider->hasClass($traitName->toString())) {
+			return [];
+		}
+
+		$errors = $this->attributesCheck->check(
+			$scope,
```
Comment (ondrejmirtes) [inline src/Rules/Traits/TraitAttributesRule.php:35]: We need to do something with $scope here. Please recreate this problem https://github.com/phpstan/phpstan-src/commit/3447391001f7a5c2bfb77f66c2d0e157242c1dae but with a trait.

We probably need to do `$scope->enterTrait()` here before calling the check. Just do `if !$scope instanceof MutatingScope then throw ShouldNotHappen`.

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]

**Comment by ondrejmirtes** [comment]
Please keep this on 2.0.x. We don't need to be adding new features to 1.12.x anymore.

File: src/Rules/Traits/TraitAttributesRule.php (hunk 11 above)
Comment (staabm) [reply src/Rules/Traits/TraitAttributesRule.php:35]: I have adjusted the PR. I think I have reproduced the problem you are after.

adding `$scope->enterTrait($classReflection)` didn't work and throws:
```
2) PHPStan\Rules\Traits\TraitAttributesRuleTest::testBug12011
PHPStan\ShouldNotHappenException: Internal error.

C:\dvl\Workspace\phpstan-src-staabm\src\Analyser\ScopeContext.php:44
C:\dvl\Workspace\phpstan-src-staabm\src\Analyser\MutatingScope.php:2935
C:\dvl\Workspace\phpstan-src-staabm\src\Rules\Traits\TraitAttributesRule.php:49
C:\dvl\Workspace\phpstan-src-staabm\src\Analyser\FileAnalyser.php:111
C:\dvl\Workspace\phpstan-src-staabm\src\Analyser\NodeScopeResolver.php:482
C:\dvl\Workspace\phpstan-src-staabm\src\Analyser\NodeScopeResolver.php:354
C:\dvl\Workspace\phpstan-src-staabm\src\Analyser\NodeScopeResolver.php:842
C:\dvl\Workspace\phpstan-src-staabm\src\Analyser\NodeScopeResolver.php:307
C:\dvl\Workspace\phpstan-src-staabm\src\Analyser\FileAnalyser.php:212
C:\dvl\Workspace\phpstan-src-staabm\src\Analyser\Analyser.php:75
C:\dvl\Workspace\phpstan-src-staabm\src\Testing\RuleTestCase.php:177
C:\dvl\Workspace\phpstan-src-staabm\src\Testing\RuleTestCase.php:138
C:\dvl\Workspace\phpstan-src-staabm\tests\PHPStan\Rules\Traits\TraitAttributesRuleTest.php:78
```

at this point I have no idea what todo. could you fix the remaining thing?

Comment (staabm) [reply src/Rules/Traits/TraitAttributesRule.php:35]: let me fix the build real quick

File: src/Rules/Traits/TraitAttributesRule.php [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+			$node->attrGroups,
+			Attribute::TARGET_CLASS,
+			'class',
+		);
+
+		$classReflection = $this->reflectionProvider->getClass($traitName->toString());
+		if (count($classReflection->getNativeReflection()->getAttributes('AllowDynamicProperties')) > 0) {
+			$errors[] = RuleErrorBuilder::message('Attribute class AllowDynamicProperties cannot be used with trait.')
+				->identifier('class.allowDynamicPropertiesTrait')
```
Comment (staabm) [inline src/Rules/Traits/TraitAttributesRule.php]: maybe?

```suggestion
				->identifier('trait.allowDynamicProperties')
```

File: src/Rules/Traits/TraitAttributesRule.php (hunk 11 above)
Comment (ondrejmirtes) [reply src/Rules/Traits/TraitAttributesRule.php:35]: Push the code that's crashing like that please, I will take a look

Comment (staabm) [reply src/Rules/Traits/TraitAttributesRule.php:35]: here you are

Comment (ondrejmirtes) [reply src/Rules/Traits/TraitAttributesRule.php:35]: Oh right, we need to be in a class first before we can enter a trait.

I'd try two things: try to put the trait reflection into enterClass (but a different ShouldNotHappen might be thrown, not sure).

If that fails, we need to come up with some fake class to put into enterClass before we can enterTrait. It might be okay to simulate some anonymous Class_ node to use.

If it's too much work, don't worry and I'll merge the original code that doesn't use MutatingScope at all.

Comment (staabm) [reply src/Rules/Traits/TraitAttributesRule.php:35]: I have class-faking without a thrown exception working. the test-case is still failling though

Will have another look after sleeping over it

**Comment by ondrejmirtes** [comment]
I changed my mind, I'm going to adjust it myself, thanks!

**Comment by ondrejmirtes** [comment]
See my commit and reasoning in description: https://github.com/phpstan/phpstan-src/pull/3738/commits/b1cebc2e1f5133b1bd68ebb4401947c5f4d0e324

**Comment by ondrejmirtes** [comment]
Please stop force-pushing here, I said I'm taking over. Thanks.

**Comment by staabm** [comment]
> Please stop force-pushing here, I said I'm taking over. Thanks.

oh sorry, did not see your comments. I got it working just a moment ago ;)

**Comment by ondrejmirtes** [comment]
Thank you!

**Comment by staabm** [comment]
thanks
